components/script/dom/document.rs

```diff
@@ -1575,6 +1612,97 @@ impl Document {
+        let document = Trusted::new(self);
+        let beforeunload_event = BeforeUnloadEvent::new(&self.window,
+                                           atom!("beforeunload"),
+                                           EventBubbles::Bubbles,
+                                           EventCancelable::Cancelable);
+        let event = beforeunload_event.upcast::<Event>();
+        event.set_trusted(true);
+        let event_status = self.window.upcast::<EventTarget>().dispatch_event_with_target(
+            document.root().upcast(),
+            &event,
+        );
+        // Step 7
+        if event.get_cancel_state() == EventDefault::Handled {
+            self.salvageable.set(false);
+        }
+        let mut can_unload = true;
+        if event_status == EventStatus::Canceled
+            || beforeunload_event.ReturnValue() == DOMString::from_string("".to_string()) {
+            // Step 8
+            // TODO: ask the user to confirm that they wish to unload the document.
+        }
+        self.decr_ignore_opens_during_unload_counter();
+        // Step 9
+        for iframe in self.iter_iframes() {
+            if let Some(document) = iframe.GetContentDocument() {
+                if !document.prompt_to_unload() {
+                    self.salvageable.set(document.salvageable());
+                    can_unload = false;
+                    break;
+                }
+            }
+        }
+        can_unload
+    }
+
+    // https://html.spec.whatwg.org/multipage/#unloading-documents
+    pub fn unload(&self) {
+        self.incr_ignore_opens_during_unload_counter();
+        let document = Trusted::new(self);
+        if self.page_showing.get() {
+            self.page_showing.set(false);
+            let event = PageTransitionEvent::new(
+                &self.window,
+                atom!("pagehide"),
+                false, // bubbles
+                false, // cancelable
+                self.salvageable.get(), // persisted
+            );
+            let event = event.upcast::<Event>();
+            event.set_trusted(true);
+            let _ = self.window.upcast::<EventTarget>().dispatch_event_with_target(
+                document.root().upcast(),
+                &event,
+            );
+        }
+        if !self.fired_unload.get() {
+            let event = Event::new(
+                &self.window.upcast(),
+                atom!("unload"),
+                EventBubbles::DoesNotBubble,
+                EventCancelable::NotCancelable,
+            );
+            event.set_trusted(true);
+            let _ = self.window.upcast::<EventTarget>().dispatch_event_with_target(
+                document.root().upcast(),
+                &event,
+            );
+            self.fired_unload.set(true);
+            let event_handled = event.get_cancel_state() == EventDefault::Handled;
+            self.salvageable.set(!event_handled);
+        }
+        // Step 13
+        for iframe in self.iter_iframes() {
```

Comment:
Like above, this will need to be done via the constellation in the case of cross origin iframes.